src/models/knn_filter_inverted_index.rs

```diff
@@ -0,0 +1,136 @@
+use crate::storage::inverted_index::{self, InvertedIndex, InvertedIndexItem};
+use std::cmp::Ordering::Equal;
+use std::cmp::Reverse;
+use std::collections::BinaryHeap;
+use std::ops::Mul;
+
+use super::lazy_load::LazyItem;
+
+pub struct ScoredCandidate(f32, u32); // (score, candidate_id)
+
+impl ScoredCandidate {
+    fn new(score: f32, candidate_id: u32) -> Self {
+        ScoredCandidate(score, candidate_id)
+    }
+}
+
+impl PartialOrd for ScoredCandidate {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.0.partial_cmp(&other.0)
+    }
+}
+
+impl Ord for ScoredCandidate {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.partial_cmp(other).unwrap()
+    }
+}
+
+impl PartialEq for ScoredCandidate {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl Eq for ScoredCandidate {}
+
+fn maintain_top_k(
+    heap: &mut BinaryHeap<Reverse<ScoredCandidate>>,
+    score: f32,
+    candidate_id: u32,
+    k: usize,
+) {
+    if heap.len() < k {
+        heap.push(Reverse(ScoredCandidate(score, candidate_id)));
+    } else if score > heap.peek().unwrap().0 .0 {
+        heap.pop();
+        heap.push(Reverse(ScoredCandidate(score, candidate_id)));
+    }
+}
+
+impl<T> InvertedIndexItem<T>
+where
+    T: Clone + 'static + Mul<Output = T> + Into<f32>,
+{
+    // Calculate distance using dot product between self and InvertedIndexItem
+    pub fn calculate_distance(&self, neighbour: &InvertedIndexItem<T>) -> f32 {
```

Comment:
We only want to find the vector_ids when traversing and not compute the distance immediately.